src/VisualStudio/IntegrationTest/TestUtilities/InProcess/ErrorList_InProc.cs

```diff
@@ -32,73 +32,145 @@ public void WaitForNoErrorsInErrorList()
-                return GetErrorListContents();
+                return GetErrorListContents(minimumSeverity);
             }
         }
 
-        private ErrorItems GetErrorItems()
-            => ((DTE2)GetDTE()).ToolWindows.ErrorList.ErrorItems;
+        private IVsEnumTaskItems GetErrorItems()
+        {
+            return InvokeOnUIThread(() =>
+            {
+                var errorList = GetGlobalService<SVsErrorList, IVsTaskList>();
+                ErrorHandler.ThrowOnFailure(errorList.EnumTaskItems(out var items));
+                return items;
+            });
+        }
     }
 
     public static class ErrorListExtensions
     {
-        public static IEnumerable<EnvDTE80.ErrorItem> AsEnumerable(this EnvDTE80.ErrorItems items)
+        public static IEnumerable<IVsTaskItem> AsEnumerable(this IVsEnumTaskItems items)
         {
-            for (var i = 1; i <= items.Count; i++)
+            var item = new IVsTaskItem[1];
+            while (true)
             {
-                yield return items.Item(i);
+                var hr = items.Next(1, item, null);
+                ErrorHandler.ThrowOnFailure(hr);
+                if (hr == VSConstants.S_FALSE)
+                {
+                    break;
+                }
+
+                // BEGIN WORKAROUND FOR https://github.com/dotnet/roslyn/issues/32121
+                // Filter out items not belonging to a currently-open project
+                if (ErrorHandler.Failed(((IVsErrorItem)item[0]).GetHierarchy(out _)))
+                {
+                    continue;
+                }
+                // END WORKAROUND FOR https://github.com/dotnet/roslyn/issues/32121
+
+                yield return item[0];
             }
         }
 
-        public static string AsString(this EnvDTE80.vsBuildErrorLevel errorLevel)
+        public static __VSERRORCATEGORY GetCategory(this IVsErrorItem errorItem)
+        {
+            ErrorHandler.ThrowOnFailure(errorItem.GetCategory(out var category));
+            return (__VSERRORCATEGORY)category;
+        }
+
+        public static string GetSeverity(this IVsTaskItem item)
+        {
+            return ((IVsErrorItem)item).GetCategory().AsString();
+        }
+
+        public static string GetDescription(this IVsTaskItem item)
+        {
+            ErrorHandler.ThrowOnFailure(item.get_Text(out var description));
+            return description;
+        }
+
+        public static string GetProject(this IVsTaskItem item)
         {
-            switch (errorLevel)
+            var errorItem = item as IVsErrorItem;
+            ErrorHandler.ThrowOnFailure(errorItem.GetHierarchy(out var hierarchy));
+            ErrorHandler.ThrowOnFailure(hierarchy.GetProperty((uint)VSConstants.VSITEMID.Root, (int)__VSHPROPID.VSHPROPID_Name, out var name));
+            return (string)name;
+        }
+
+        public static string GetFileName(this IVsTaskItem item)
+        {
+            ErrorHandler.ThrowOnFailure(item.Document(out var fileName));
```

Comment:
Does this return an empty string for error list items without document name?

Reply:
I'm not sure. It's the pattern I used in the test harness experiment and it worked for all the tests we have today. We can always adjust it if we need to test a new scenario _and_ we haven't completed the switch to Apex yet.